contracts/cw20-merkle-airdrop/src/msg.rs

```diff
@@ -0,0 +1,64 @@
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+use cosmwasm_std::Uint128;
+
+#[derive(Serialize, Deserialize, JsonSchema)]
+pub struct InstantiateMsg {
+    pub owner: String,
+    pub cw20_token_address: String,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum ExecuteMsg {
+    UpdateConfig {
+        owner: Option<String>,
+    },
+    RegisterMerkleRoot {
+        merkle_root: String,
+    },
+    Claim {
+        stage: u8,
+        amount: Uint128,
+        proof: Vec<String>,
+    },
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum QueryMsg {
+    Config {},
+    MerkleRoot { stage: u8 },
+    LatestStage {},
+    IsClaimed { stage: u8, address: String },
+}
+
+#[derive(Serialize, Deserialize, Clone, PartialEq, JsonSchema, Debug)]
+#[serde(rename_all = "snake_case")]
+pub struct ConfigResponse {
+    pub owner: String,
+    pub cw20_token_address: String,
+}
+
+// We define a custom struct for each query response
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct MerkleRootResponse {
+    pub stage: u8,
+    pub merkle_root: String,
```

Comment:
I see, that sending hashes as `String` in message (in particular json) seems to be simple and clean, but the issue I have with that, is that it makes message bigger, and also requires additional parsing - all sums up to unnecessary gas cost increase. Also it increases class of invalid messages. I wonder if it isn't better to use `[u8; 32]` as this type (and also the same for `proof`). I guess that at the end noone would write those things by hand, so basically it should be no problem to have conversion from human-readable hash to just array of bytes.

However - I don't have strong opinion on this. @ethanfrey what is approach we prefer in such cases?

Reply:
I mention above the issues, since we are limited to JSON parsing here, not internal structs, or binary protocol.
All input-output over Msg or Storage layer will be inefficient as JSON is nowhere like writing raw bytes.

I prefer using `Binary` in general, which we can use like Vec<u8> (or &[u8]) in the contract and transparently handles base64 encoding/decoding.

If you want hex representation, so people can easier inspect it, Orkun has the write approach. We could also add a helper to cosmwasm-std HexBytes, which is like Binary but using a hex-encoding. Not sure there are so many cases, but people do seem to like to compare hashes in hex.